pkg/providers/secretmanager/secretmanager.go

```diff
@@ -0,0 +1,87 @@
+package secretmanager
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strings"
+
+	"gopkg.in/yaml.v3"
+
+	smpb "google.golang.org/genproto/googleapis/cloud/secretmanager/v1beta1"
+
+	sm "cloud.google.com/go/secretmanager/apiv1beta1"
+	"github.com/variantdev/vals/pkg/api"
+)
+
+// Format: ref+secretmanager://project/mykey[?version=VERSION]#/yaml_or_json_key/in/secret
+type provider struct {
+	client  *sm.Client
+	ctx     context.Context
+	version string
+}
+
+func New(cfg api.StaticConfig) *provider {
+	ctx := context.Background()
+
+	p := &provider{
+		ctx: ctx,
+	}
+
+	version := cfg.String("version")
+	if version == "" {
+		p.version = "latest"
+	} else {
+		p.version = version
+	}
+
+	return p
+}
+
+func (p *provider) GetString(key string) (string, error) {
+
+	secret, err := p.getSecretBytes(key)
+	if err != nil {
+		return "", err
+	}
+
+	return string(secret), nil
+}
+
+func (p *provider) GetStringMap(key string) (map[string]interface{}, error) {
```

Comment:
This function is used by the vals engine when you included `#/PATH/TO/VALUE` in your ref URL.

You usually use it when the secrets manager has a kind of nested secrets.

A few examples:

- AWS SSM Parameter Store supports querying a set of parameters in one-shot. `ref+awssm://foo/bar#/a` results in querying all the parameters whose keys are prefixed with `foo/bar`. After that `#/a` extracts the value for `foo/bar/a` from it. Later you can use `ref+awsssm://foo/bar#/b` which reuses the previously cached result of `foo/bar` to return `foo/bar/b`.
- AWS doesn't support prefix searches like SSM ParameterStore. But it has a convention that each secret value can be a structured document(like JSON). So, we usually use `ref+awssecrets://foo/bar#/a` to get the secret value named `foo/bar`. The secret value contains a JSON object from where we extract the value for key `.a`.

So the question here is - does GCP Secrets Manager has a convention like AWS SecretsManager, or a dedicated API for querying a set of secrets in one API call? If neither applies, you can skip implementing this, like I've done for e.g. the tfstate provider: https://github.com/variantdev/vals/blob/master/pkg/providers/tfstate/tfstate.go#L39

Reply:
> dedicated API for querying a set of secrets in one API call?

Not to my knowledge

> has a convention that each secret value can be a structured document

Yes, I expect that to be the case.  The cost per secret would make it silly to have each key for each app be a standalone secret.  I would expect an application to want to use a single secret to store all values as a json/yaml blob in most cases.  

As implemented, I support this:

```
% gcloud secrets versions access latest --secret valstestvar
foo: bar
% echo 'testval: ref+secretmanager://secret-test-272302/valstestvar#/foo' | ./bin/vals eval
testval: bar
% echo 'testval: ref+secretmanager://secret-test-272302/valstestvar' | ./bin/vals eval
testval: |
    foo: bar
```